### ISSUE TITLE
[API Proposal]: Adding a Cardinality/PopCount property to BitArray

### ISSUE THREAD

**Issue by polijp** [dotnet/runtime#104299]
### Background and motivation

You often need to know how many bits are set to 1 in a BitArray.
Presently, there are two ways to achieve that:
- either you copy the BitArray instance to an int[] and then you count (eventually using intrinsics)
- or you access one by one the bits.

Adding a Cardinality property takes the advantage of the underlying int array. It could use the instrinsic BitOperations.PopCount() method.

### API Proposal

```csharp
namespace System.Collections;

public class BitArray ...
{
    public int PopCount => ...
}
```

### API Usage

```csharp
// Fancy the value
var c = new BitArray(10);
if (c.PopCount %2 ==0) do something;



### Alternative Designs

_No response_

### Risks

_No response_

**Comment by sebastienros** [comment]
Also discussed in https://github.com/dotnet/runtime/issues/84907

**Comment by tannergooding** [comment]
This would need to be a `int GetPopCount()` or `static int PopCount(BitArray bitArray)` API.

The API has to iterate all entries in the underlying storage for the `BitArray` and so would not qualify, under the Framework Design Guidelines, to be a property. The static method would be the most similar to how the `PopCount` API is exposed for other types and would be my preference, accordingly.

Happy to take this to API review if the proposal is updated.

**Comment by LeaFrock** [comment]
How about the following?

```diff

public class BitArray 
{
    public bool HasAllSet();

    public bool HasAnySet();

+    public int CountCurrentSet();
}

```

**Comment by tannergooding** [comment]
The name we use elsewhere is `PopCount`, we likely shouldn't deviate. We likewise shouldn't expose both a static and an instance method. Given other pop count methods tend to be static, I'd suggest that this one be as well.

Either the original proposer needs to update their post or a new one needs to be created as well.

**Comment by LeaFrock** [comment]
If we focus on the BitArray itself, `CountCurrentSet` probably belongs to the same “family” as the `HasXXXSet` methods, even though at the lowest level it’s essentially a pop-count operation. For most application developers who don’t use bit-level operations every day, that framing makes the API’s purpose easier to grasp. Likewise, I believe an instance method is preferable to a static one, keeping it consistent with the existing `XXXSet` APIs.

> Either the original proposer needs to update their post or a new one needs to be created as well.

I'll wait for the update several weeks and create another issue if the original proposer does not.

**Comment by tannergooding** [comment]
API review is unlikely to approve naming deviations because `PopCount` is the standard name that exists throughout. We already had the discussion on `CountSetBits` or the like and settled on the former because its a fairly industry standard term. It's even exposed as `int.PopCount(x)`.

Similar reasoning exists for it being a static method, consistency with other types and APIs exposed.

**Comment by LeaFrock** [comment]
> We already had the discussion on CountSetBits ...

Sorry I don't find the discussion you mention in this repo. Anyway, hope the runtime ship it and move on.

@polijp Would you mind update the proposal?

**Comment by tannergooding** [comment]
> Sorry I don't find the discussion you mention in this repo

It was had in API review when `BitOperations.PopCount` was exposed. This was years ago at this point and I don't have a link to the exact recording on YouTube that would have it anymore. -- A large factor of this decision was that it is the name of the intstruction used by Arm64, x64, and WASM; as well the name used in C++ and several other ecosystems. Not all of them do of course, Java uses `countBits` and Rust uses `count_ones` for example; but `popcount` is by far the predominant name.

This decision was then reaffirmed in .NET 7 when we exposed `T.PopCount` as part of Generic Math and made it directly exposed as part of all primitive integral types.

**Comment by polijp** [comment]
Hi everyone.
So I updated the proposal. If I summarize the discussion:
- name should be PopCount
- it cannot be a property (because it has to go through the internal array)
- must be static to be homogeneous with other PopCount.

There is a point that is still open: either it should be a static or instance method.

I’d love to see it dotnet 10. :)

**Comment by tannergooding** [comment]
It should be static for the reasons I've stated above and consistency with the rest of the .NET ecosystem in how we expose `PopCount` on types.

API review is likely not to approve the instance method.

**Comment by polijp** [comment]
Ok updated.
Do you think it’s ok to add an overload to BitOperations.PopCount(…) pls ?

**Comment by tannergooding** [comment]
> Do you think it’s ok to add an overload to BitOperations.PopCount(…) pls ?

We largely aren't versioning `BitOperations` anymore, not since Generic Math was introduced in .NET 7. What kind of API are you looking for here?

We wouldn't want to add a dependency between `BitOperations` and `BitArray` as they exist at different layers.

**Comment by LeaFrock** [comment]
> I’d love to see it dotnet 10. :)

Is there still time 😆 ?

**Comment by tannergooding** [comment]
> Is there still time 😆 ?

No. It's too late for .NET 10 as we've already snapped for RC1 and main is now targeting .NET 11

You can use `CollectionsMarshal.AsBytes(BitArray array)` + `TensorPrimitives.PopCount` to fill the gap in the interim.

**Comment by bartonjs** [comment]
[Video](https://www.youtube.com/watch?v=5rKlRoQHCVU&t=1h13m31s)

* While PopCount is static everywhere else, it should be instance here to be consistent with other BitArray members
* We should add lzcount and tzcount at the same time

```C#
namespace System.Collections;

public partial class BitArray
{
    public int PopCount();
    public int LeadingZeroCount();
    public int TrailingZeroCount();
}
```

**Comment by IS4Code** [comment]
I must comment on the very unfortunate choice of naming, considering `PopCount` is not popping anything. I presume many people will avoid this method considering its name suggests mutation. I don't think consistency with obscure naming schemes should trump .NET's own conventions, and "verb + object" is at the top for method naming. We wouldn't want to have horrible things like `wcsrtombs_s` in .NET either.
